packages/store/src/StoreCore.sol

```diff
@@ -0,0 +1,559 @@
+    for (uint8 i; i < schema.numDynamicFields(); ) {
+      dynamicDataLocation = StoreCoreInternal._getDynamicDataLocation(table, key, i);
+      dynamicDataLength = dynamicLength.atIndex(i);
+      Storage.store({
+        storagePointer: dynamicDataLocation,
+        offset: 0,
+        memoryPointer: memoryPointer,
+        length: dynamicDataLength
+      });
+      memoryPointer += dynamicDataLength; // move the memory pointer to the start of the next dynamic data
+      unchecked {
+        i++;
+      }
+    }
+
+    // Emit event to notify indexers
+    emit MudStoreSetRecord(table, key, data);
+  }
+
+  function setField(
+    bytes32 table,
+    bytes32[] memory key,
+    uint8 schemaIndex,
+    bytes memory data
+  ) internal {
+    Schema schema = getSchema(table);
+
+    // Call onSetField hooks (before actually modifying the state, so observers have access to the previous state if needed)
+    address[] memory hooks = HooksTable.get(table);
+    for (uint256 i = 0; i < hooks.length; i++) {
+      IStoreHook hook = IStoreHook(hooks[i]);
+      hook.onSetField(table, key, schemaIndex, data);
+    }
+
+    if (schemaIndex < schema.numStaticFields()) {
+      StoreCoreInternal._setStaticField(table, key, schema, schemaIndex, data);
+    } else {
+      StoreCoreInternal._setDynamicField(table, key, schema, schemaIndex, data);
+    }
+
+    // Emit event to notify indexers
+    emit MudStoreSetField(table, key, schemaIndex, data);
+  }
+
+  function deleteRecord(bytes32 table, bytes32[] memory key) internal {
+    // Get schema for this table
+    Schema schema = getSchema(table);
+
+    // Call onDeleteRecord hooks (before actually modifying the state, so observers have access to the previous state if needed)
+    address[] memory hooks = HooksTable.get(table);
+    for (uint256 i = 0; i < hooks.length; i++) {
+      IStoreHook hook = IStoreHook(hooks[i]);
+      hook.onDeleteRecord(table, key);
+    }
+
+    // Delete static data
+    uint256 staticDataLocation = StoreCoreInternal._getStaticDataLocation(table, key);
+    Storage.store({ storagePointer: staticDataLocation, offset: 0, data: new bytes(schema.staticDataLength()) });
+
+    // If there are no dynamic fields, we're done
+    if (schema.numDynamicFields() == 0) return;
+
+    // Delete dynamic data
+    uint256 dynamicDataLocation;
+    PackedCounter encodedLengths = StoreCoreInternal._loadEncodedDynamicDataLength(table, key);
+    for (uint8 i; i < schema.numDynamicFields(); ) {
+      dynamicDataLocation = StoreCoreInternal._getDynamicDataLocation(table, key, i);
+      Storage.store({ storagePointer: dynamicDataLocation, offset: 0, data: new bytes(encodedLengths.atIndex(i)) });
+      unchecked {
+        i++;
+      }
+    }
+
+    // Delete dynamic data length
+    uint256 dynamicDataLengthLocation = StoreCoreInternal._getDynamicDataLengthLocation(table, key);
+    Storage.store({ storagePointer: dynamicDataLengthLocation, data: bytes32(0) });
```

Comment:
Deleting dynamic data seems unnecessary, since you delete its length; a new record that sets the new length would be responsible for initializing the elements